File: leetcode/binary-tree-postorder-traversal/binary-tree-postorder-traversal.cpp
/**************************************
* @author dujiong
* @date 2016.9.26
* @version V0.1
**************************************/

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode *root) {
        vector<int> ret;
        if(root==NULL){
            return ret;
        }
        stack<TreeNode*> treeStack;
        treeStack.push(root);
        TreeNode *temp = NULL;
        
        while(!treeStack.empty()){
            temp = treeStack.top();
            if(temp->left==NULL && temp->right==NULL){
                ret.push_back(temp->val);
                treeStack.pop();
            }else{
                if(temp->right){
                    treeStack.push(temp->right);
                    temp->right = NULL;
                }
                if(temp->left){
                    treeStack.push(temp->left);
                    temp->left = NULL;
                }
            }
        }
        return ret;
    }
};

File: leetcode-golang/Binary Tree Postorder Traversal/Binary Tree Postorder Traversal.go
type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

//非递归的方式
func postorderTraversal(root *TreeNode) []int {
    ret := []int{}
    if root == nil {
        return ret
    }
    nodes := []*TreeNode{}
    var r *TreeNode = nil
    for len(nodes)>0 || root != nil {
        if root != nil {
            nodes = append(nodes, root)
            root = root.Left
        } else {
            root = nodes[len(nodes)-1]
            if root.Right != nil && root.Right != r {
                root = root.Right
                nodes = append(nodes, root)
                root = root.Left
            } else {
                ret = append(ret, root.Val)
                nodes = nodes[:len(nodes)-1]
                r = root
                root = nil
            }
        }
    }
    return ret
}

//递归的方式
func postorderTraversal(root *TreeNode) []int {
	var result []int
	if root == nil {
		return result
	}
	result = append(result, postorderTraversal(root.Left)...)
	result = append(result, postorderTraversal(root.Right)...)
	result = append(result, root.Val)
	return result
}


